Adding long Press support on android and iOS

shubhamguptadream11 opened the pull request:
Solve this: https://github.com/okwasniewski/react-native-bottom-tabs/issues/58

**Summary**:
This PR introduces the handling of long press events on Android for tabs in the ReactBottomNavigationView. While Android's BottomNavigationView does not natively support long press events as a first-class feature (like setOnItemSelectedListener for selection), we have implemented custom logic to detect and handle long presses on individual tabs.

**Implementation:**

**Android**

**Long Press Handling:** We attach OnLongClickListener to each tab using setOnLongClickListener, triggering a callback when a tab is long-pressed.

We use the post {} function to defer the registration of long press listeners until after the view hierarchy has been fully laid out. This ensures that all views for the tabs are fully initialised before attaching the listeners. Attempting to set the listeners earlier may lead to issues, as the views may not yet be available.

**iOS**

- **TabItemLongPressModifier**: Uses SwiftUIIntrospect to introspect the TabView and add a long press gesture to the tab bar.
- **LongPressGestureHandler**: A custom gesture handler class that processes the long press gesture and identifies the index of the pressed tab based on touch location.
- **SwiftUIIntrospect** **Dependency**: Added "SwiftUIIntrospect", '~> 1.0' to the podspec to enable the use of introspect.


**RN API**

**Long Press Event Callback for Tabs**
With this PR, a new tabLongPress event is introduced for handling long press actions on tabs in Android's BottomNavigationView. This allows developers to subscribe to long press events for both active and inactive tabs, providing more flexibility in user interactions.

When a user long-presses a tab (including the currently active tab), the tabLongPress callback, if subscribed via screenOptions, will be triggered, enabling custom actions in the navigation logic.

```
screenListeners={{
  tabLongPress: (data) => {
    console.log(
      '### JS Navigator level TabLongPress called',
       data
    );
  },
}}
```

**Testing Plan:**

Tested the integration in these 3 ways.
```
    <Tab.Navigator
      screenListeners={{
        tabLongPress: (data) => {
          console.log(
            '### JS Navigator level TabLongPress called',
            JSON.stringify(data)
          );
        },
      }}
    >
      <Tab.Screen
        name="Article"
        component={Article}
        listeners={{
          tabLongPress: (data) => {
            console.log(
              '### JS Tab level TabLongPress called',
              JSON.stringify(data)
            );
          }
        }}
        options={{
          tabBarBadge: '10',
          tabBarIcon: ({ focused }) =>
            focused
              ? require('../../assets/icons/person_dark.png')
              : require('../../assets/icons/article_dark.png'),
        }}
      />
```
- By adding tabLongPress in screenListeners at Navigator level
- By adding tabLongPress in listeners at individual tab level
- By adding addListener inside all screens like Article screen.

okwasniewski commented on android/src/main/java/com/rcttabview/RCTTabView.kt:
```diff
@@ -45,6 +48,26 @@ class ReactBottomNavigationView(context: Context) : BottomNavigationView(context
       onTabSelected(item)
       true
     }
+    // Register long press listener for each tab
+    post {
+      for (i in 0 until menu.size()) {
+        val menuItem = menu.getItem(i)
+        findViewById<View>(menuItem.itemId).setOnLongClickListener {
+          onTabLongPressed(menuItem)
+          true
+        }
+      }
+    }
+  }
```
I wonder what happens if you dynamically add tabs, this probably won't work for the newly added tab. Maybe this can be moved to `items` setter.

shubhamguptadream11 replied:
Yeah, will move this block to updateItems function. There it will handle dynamic tabs case as well.

okwasniewski commented on example/src/Examples/NativeBottomTabs.tsx:
```diff
@@ -9,10 +9,27 @@ const Tab = createNativeBottomTabNavigator();
   return (
-    <Tab.Navigator>
+    <Tab.Navigator
+      screenListeners={{
+        tabLongPress: (data) => {
+          console.log(
+            '### JS Navigator level TabLongPress called',
+            JSON.stringify(data)
+          );
+        },
+      }}
+    >
```
Could you test the integration like this? (From JS bottom tabs docs)

```
React.useEffect(() => {
  const unsubscribe = navigation.addListener('tabLongPress', (e) => {
    // Do something
  });

  return unsubscribe;
}, [navigation]);
```

shubhamguptadream11 replied:
Yeah sure!!

shubhamguptadream11 replied:
@okwasniewski 
I was trying to test this as suggested by you?
But I found that even **tabPress** is not working in this way.

```
import { useNavigation } from '@react-navigation/native';

React.useEffect(() => {
  const unsubscribe = navigation.addListener('tabPress', (e) => {
    // Do something
  });

  return unsubscribe;
}, [navigation]);
```

Is this expected?

okwasniewski replied:
Great! Can you use it inside of the test app?

shubhamguptadream11 replied:
Do you mean example app?

okwasniewski replied:
Yes, to replace 

```
screenListeners={{
        tabLongPress: (data) => {
```

With the listeners inside of screens

okwasniewski reviewed (commented):
Thanks for working on this. It's definitely moving in the right direction so you can work on something similar for iOS 👍🏻

okwasniewski commented on ios/TabViewImpl.swift:
```diff
@@ -59,6 +60,12 @@ struct TabViewImpl: View {
               labeled: props.labeled
             )
           }
+          .onLongPressGesture {
```
You should attach `.onLongPressGesture` to individual tab items

shubhamguptadream11 replied:
I tried like this as well but didn't worked.

```
            TabItem(
              title: tabData?.title,
              icon: icon,
              sfSymbol: tabData?.sfSymbol,
              labeled: props.labeled
            )
            .onLongPressGesture {
                print("### Shubham iOS Side onLongPressGesture called")
                if let key = tabData?.key {
                    onLongPress(key)
                }
            }
```

okwasniewski replied:
I think you need to attach it inside of tabItem(?) To label 

```swift
 .tabItem {
                Label("Home", systemImage: "house")
                    .onLongPressGesture {
                        // Something
                    }
            }
```

Does it work this way?

shubhamguptadream11 replied:
No, this is also not working.

okwasniewski replied:
It looks like its totally broken 😅

I found this: https://procomputingproducts.co.uk/post/how-to-detect-a-long-press-gesture-on-a-tab-item-label

But it didn't work for me too

shubhamguptadream11 replied:
Yeah I also tried this. Not working for me as well. What do you think should be the next steps?

okwasniewski replied:
I got this working. 

It's really verbose solution but I don't have anything better. 


1. We need to add SwiftUI Introspect as dependency

In podspec

```
  s.dependency 'SwiftUIIntrospect', '~> 1.0'
```

2. Override gesture recognizer of tab view (Let's put this in a separate file): 


```swift
import SwiftUIIntrospect

struct TabItemLongPressModifier: ViewModifier {
  let onLongPress: (Int) -> Void
  
  func body(content: Content) -> some View {
    content.introspect(.tabView, on: .iOS(.v13, .v14, .v15, .v16, .v17, .v18)) { tabController in
      // Remove existing long press gestures
      if let existingGestures = tabController.tabBar.gestureRecognizers {
        for gesture in existingGestures where gesture is UILongPressGestureRecognizer {
          tabController.tabBar.removeGestureRecognizer(gesture)
        }
      }
      
      // Create gesture handler
      let handler = LongPressGestureHandler(tabBar: tabController.tabBar, handler: onLongPress)
      let gesture = UILongPressGestureRecognizer(target: handler, action: #selector(LongPressGestureHandler.handleLongPress(_:)))
      gesture.minimumPressDuration = 0.5
      
      objc_setAssociatedObject(tabController.tabBar, &AssociatedKeys.gestureHandler, handler, .OBJC_ASSOCIATION_RETAIN)
      
      tabController.tabBar.addGestureRecognizer(gesture)
    }
  }
}

private struct AssociatedKeys {
  static var gestureHandler: UInt8 = 0
}

private class LongPressGestureHandler: NSObject {
  private weak var tabBar: UITabBar?
  private let handler: (Int) -> Void
  
  init(tabBar: UITabBar, handler: @escaping (Int) -> Void) {
    self.tabBar = tabBar
    self.handler = handler
    super.init()
  }
  
  @objc func handleLongPress(_ recognizer: UILongPressGestureRecognizer) {
    guard recognizer.state == .began,
          let tabBar = tabBar else { return }
    
    let location = recognizer.location(in: tabBar)

    // Get buttons and sort them by frames 
    let tabBarButtons = tabBar.subviews.filter { String(describing: type(of: $0)).contains("UITabBarButton") }.sorted(by: {$0.frame.minX < $1.frame.minX})
    
    for (index, button) in tabBarButtons.enumerated() {
      if button.frame.contains(location) {
        handler(index)
        break
      }
    }
  }
}

extension View {
  func onTabItemLongPress(_ handler: @escaping (Int) -> Void) -> some View {
    modifier(TabItemLongPressModifier(onLongPress: handler))
  }

```

Then we can add the modifier: 

```swift
TabView(selection: $props.selectedPage) {
      
    }
    .onTabItemLongPress({ index in
      print("LONG PRESS DECTEDTED: ", index)
    })
```

I'll try to look for a better option but I didn't find anything better. We can also make it better in the future

shubhamguptadream11 replied:
One thing which I want to point out is, LongPressGesture is working fine when attaching it on hostViewController.

```
       let longPressGesture = UILongPressGestureRecognizer(target: self, action: #selector(handleLongPress(_:)))
       hostingController?.view.addGestureRecognizer(longPressGesture)
```
in `setupView` function.

okwasniewski replied:
Yeah, I noticed it too. .longPressGesture is also working on the RepresentableView but not on tabBarItem. 

I think this might be a bug in SwiftUI

shubhamguptadream11 commented:
@okwasniewski Android & iOS implementation are added and tested. You can start reviewing this PR.
Thanks for all your help😌

okwasniewski reviewed (commented):
Thanks for working on this! 

Can you also add section about `Events` to the docs?

https://reactnavigation.org/docs/bottom-tab-navigator#events

Also I noticed we are missing long press example, can you add it?

shubhamguptadream11 commented:
@okwasniewski 
Following changes are done post your review:
- Added **events** section in doc and added `tabLongPress`.
- Incorporated API related changes regarding index in `onTabLongPress`.
- Added example in `NativeBottomTab` itself.

Thanks again for your feedback!!

okwasniewski reviewed (approved):
Thank you for working on this! 🚀
